uSDHC Hotplugging

midnightveil opened the pull request:
- Virtualiser reading the MBR and assigning partitions with a static partition index <-> client index mapping is split out into a separate virtualiser policy interface, implemented by mbr.c (& virt.h)
- Added a hotplugging example that plays a "game" with a "safe" ejection state (see its README)
- ~~Split out `blk_storage_info_t` to a `storage_info.h` header with some helper functions~~
- New interclient hotplug interface (sddf/hotplug/clients.h) and driver control interface (sddf/hotplug/control.h)
- uSDHC driver polls GPIO pins for card detection

To-Do:
- [x] ~~Design document (in progress)~~ added to design doc
- [ ] use gpio driver when it exists
- [ ] integration with sdfgen tooling

midnightveil committed:
blk/mmc: Read GPIO registers for card detect (TEMP DO NOT MERGE)

This should be replaced by a proper GPIO driver.

Signed-off-by: julia <git.ts@trainwit.ch>

midnightveil committed:
blk/mmc: Add hotplug control requests support

This allows clients to request for the device to be inserted or ejected.

Signed-off-by: julia <git.ts@trainwit.ch>

midnightveil committed:
hotplug: Add an inter-client hotplug interface

Signed-off-by: julia <git.ts@trainwit.ch>

Ivan-Velickovic commented on include/sddf/blk/storage_info.h:
When you get a chance do you mind putting this in another PR so we can merge it before looking at the hot-plugging stuff? It would be nice to have `blk_storage_is_ready` etc available.

midnightveil committed:
blk/mmc: hotplugging via ready state notifications

- Makes the driver able to go from ready->idle and idle->ready
- Makes the virt able to do the same
- Also adds a very basic GPIO driver with polling, this should be replaced
  with a proper GPIO driver. This is to test card hotplugging.

midnightveil committed:
examples/mmc_hotplug: (BROKEN!) old mmc hotplug example
